packages/liveblocks-core/src/connection.ts

```diff
@@ -724,29 +767,29 @@ function createConnectionStateMachine<T extends BaseAuthResult>(
         if (e.event.code === 4999) {
           return {
             target: "@idle.failed",
-            effect: log(
-              LogLevel.WARN,
-              "Connection to WebSocket closed permanently. Won't retry."
-            ),
+            effect: logPermanentClose,
           }; // Should not retry, give up
         }
 
+        // Server instructs us to reauthenticate
+        if (e.event.code === 4001 /* Not Allowed */) {
+          return {
+            target: "@auth.backoff",
+            effect: [increaseBackoffDelay, logCloseEvent(e.event)],
+          };
+        }
```

Comment:
This is an addition that fixes the bug. All WS status codes just retry connecting (with an aggressive backoff), but 4001 specifically is an instruction from the server to get a new token, because the current one is not authorized explicitly. Note that in this case we transition to `@auth.backoff` instead of `@connecting.backoff`, which we do in all other cases.